x/uniswap/internal/types/msgs.go

```diff
@@ -0,0 +1,230 @@
+package types
+
+import (
+	"strings"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+var (
+	_ sdk.Msg = MsgSwapOrder{}
+	_ sdk.Msg = MsgAddLiquidity{}
+	_ sdk.Msg = MsgRemoveLiquidity{}
+)
+
+/* --------------------------------------------------------------------------- */
+// MsgSwapOrder
+/* --------------------------------------------------------------------------- */
+
+// MsgSwap Order - struct for swapping a coin
+type MsgSwapOrder struct {
+	SwapDenom  string         `json:"swap_denom"`   // The desired denomination either to be bought or sold
+	Amount     sdk.Coins      `json:"amount"`       // The specified amount to be either bought or sold
+	Bound      sdk.Int        `json:"bound"`        // If buy order, maximum amount of coins to be sold; otherwise minimum amount of coins to be bought
+	Deadline   time.Time      `json:"deadline"`     // deadline for the transaction to still be considered valid
+	Sender     sdk.AccAddress `json:"sender"`       // address swapping coin
+	Recipient  sdk.AccAddress `json:"recipient"`    // address output coin is being sent to
+	IsBuyOrder bool           `json:"is_buy_order"` // boolean indicating whether the order should be treated as a buy or sell
+}
+
+// NewMsgSwapOrder is a constructor function for MsgSwapOrder
+func NewMsgSwapOrder(
+	swapDenom string, amt sdk.Coins, bound sdk.Int, deadline time.Time,
+	sender, recipient sdk.AccAddress, isBuyOrder bool,
+) MsgSwapOrder {
+
+	return MsgSwapOrder{
+		SwapDenom:  swapDenom,
+		Amount:     amt,
+		Bound:      bound,
+		Deadline:   deadline,
+		Sender:     sender,
+		Recipient:  recipient,
+		IsBuyOrder: isBuyOrder,
+	}
+}
+
+// Route Implements Msg
+func (msg MsgSwapOrder) Route() string { return RouterKey }
+
+// Type Implements Msg
+func (msg MsgSwapOrder) Type() string { return "swap_order" }
+
+// ValidateBasic Implements Msg.
+func (msg MsgSwapOrder) ValidateBasic() sdk.Error {
+	if strings.TrimSpace(msg.SwapDenom) == "" {
+		return ErrNoDenom(DefaultCodespace)
+	}
+	// initially only support trading 1 coin only
+	if len(msg.Amount) != 1 {
```

Comment:
I was thinking of possibly allowing buy/sell order with multi coin amounts. For example: if someone wanted to trade all their eth, btc, etc for atoms. Then it could be done in one msg. Not sure how useful it would be, this could also be an upgrade that is done later on

Reply:
nevermind, will be easy enough to do multimsg transaction